python/lsst/daf/butler/cli/utils.py

```diff
@@ -440,3 +567,259 @@ def make_metavar(self):
+        if cls._name is None:
+            p = click.Option(cls.optionFlags())
+            cls._name = p.name
+            cls._opts = p.opts
+
+    @classmethod
+    def optionKey(cls):
+        """Get the keyword argument name for an option flag."""
+        cls._update()
+        return cls._name
+
+    @classmethod
+    def flag(cls):
+        """Get the option flag. Prefers the first flag that begins with two
+        dashes if one is declared, otherwise uses the first flag that begins
+        with a single dash."""
+        cls._update()
+        ret = None
+        for opt in cls._opts:
+            if ret is None or (ret.startswith("-") and opt.startswith("--")):
+                ret = opt
+            if ret.startswith("--"):
+                break
+        return ret
+
+
+class MWCommand(click.Command):
+    """Command subclass that stores a copy of the args list for use by the
+    command."""
+
+    def parse_args(self, ctx, args):
+        MWCtxObj.getFrom(ctx).args = copy.copy(args)
+        super().parse_args(ctx, args)
+
+
+class MWCtxObj():
+    """Helper object for managing the `click.Context.obj` parameter, allows
+    obj data to be managed in a consistent way.
+
+    `Context.obj` defaults to None. `MWCtxObj.getFrom(ctx)` can be used to
+    initialize the obj if needed and return a new or existing MWCtxObj.
+
+    Attributes
+    ----------
+    args : `list` [`str`]
+        The list of arguments (argument values, option flags, and option
+        values), split using whitespace, that were passed in on the command
+        line for the subcommand represented by the parent context object.
+    """
+
+    def __init__(self):
+
+        self.args = None
+
+    @staticmethod
+    def getFrom(ctx):
+        """If needed, initialize `ctx.obj` with a new MWCtxObj, and return the
+        new or already existing MWCtxObj."""
+        if ctx.obj is not None:
+            return ctx.obj
+        ctx.obj = MWCtxObj()
+        return ctx.obj
+
+
+class ForwardOptions:
+    """Captures CLI options to be forwarded from one subcommand to future
+    subcommands executed as part of a single CLI command invocation
+    (called "chained commands").
+
+    Parameters
+    ----------
+    optionDecorators : `list` or `tuple` [`MWOptionDecorator`]
+        Represents the options that should be forwarded from one subcommand to
+        the next.
+
+    Attributes
+    ----------
+    functionKwargs : `dict` [`str`, `str`]
+        The cached kwargs (argument name and argument value) from subcommands
+        that were called with values passed in as options on the command line.
+    """
+
+    def __init__(self, mwOptions=None):
+        self.functionKwargs = {}
+
+    @staticmethod
+    def _getKwargsToSave(mwOptions, arguments, **kwargs):
+        """Get kwargs that should be cached for use by subcommands invoked in
+        the future.
+
+        Parameters
+        ----------
+        mwOptions : `list` or `tuple` [`MWOption`]
```

Comment:
Can be `iterable` instead of `list or tuple`?

Reply:
well I guess it could be a dict but it really oughtn't be.